src/hist1d.jl

```diff
@@ -35,20 +52,59 @@ function bincenters(h::Hist1D)
 end
 
 """
-    push!(h::Hist1D, val::Real, wgt::Real=one{T})
+    empty!(h::Hist1D)
+
+Resets a histogram's bin counts and `sumw2`.
+"""
+function Base.empty!(h::Hist1D{T,E}) where {T,E}
+    h.hist.weights .= zero(T)
+    h.sumw2 .= 0.0
+    return h
+end
+
+"""
+    unsafe_push!(h::Hist1D, val::Real, wgt::Real=1)
+    push!(h::Hist1D, val::Real, wgt::Real=1)
 
 Adding one value at a time into histogram. 
 `sumw2` (sum of weights^2) accumulates `wgt^2` with a default weight of 1.
+`unsafe_push!` is a faster version of `push!` that is not thread-safe.
+
+N.B. To append multiple values at once, use broadcasting via
+`push!.(h, [-3.0, -2.9, -2.8])` or `push!.(h, [-3.0, -2.9, -2.8], 2.0)`
 """
-function Base.push!(h::Hist1D{T,E}, val::Real, wgt::Real=1.0) where {T,E}
-    @inbounds binidx = searchsortedlast(h.hist.edges[1], val)
+@inline function Base.push!(h::Hist1D{T,E}, val::Real, wgt::Real=1) where {T,E}
+    r = h.hist.edges[1]
+    L = length(r) - 1
+    start = first(r)
+    stop = last(r)
+    c = ifelse(val > stop, 0, 1)
```

Comment:
actually, because we're using a `lock` here, it might be better to not `lock` it in case overflow underflow.

Of course this depends on how often we over/underflow and also what's our design gonna be for actual over/underflow

Reply:
So, keep the current implementation until after working out over/underflow ideas?

Reply:
yeah. if we really want we can split a third `unsafe_push!(H, binidx, count)` which contains only the "unsafe" part and lock around this.... details :(  I doubt would matter too much, let's get the functionality flushed out first